File: Chapter1/S1.4/R/exercise07.r
course <- c(rep("superior", 2), rep("medio", 6), rep("fundamental", 2))
age <- c(34, 43, 31, 37, 24, 25, 27, 22, 21, 26)
salary <- c(1100.0, 1450.0, 960.0, 960.0, rep(600.0, 3), rep (450.0, 3))
jobYears <- c(5, 8, 6, 8, 3, 2, 5, 2, 3, 3)

df <- data.frame(course, age, salary, jobYears)

# Course variable
course_freq <- table(df$course)
print("Histogram for course")
print(course_freq)

# Get percentage values for each categorie
pct <- round(course_freq/sum(course_freq)*100)
# Place percentage value with the label names
lbls <- paste(rownames(course_freq), pct)
# Place a % symbol to the label text
lbls <- paste(lbls, "%", sep="")
pie(course_freq, labels=lbls, main="Course variable")

# Group Age variable
age_bins = pd.cut(df["Age"], (20, 25, 30, 35, 40, 45), \
                  labels=["(20, 25]", "(25, 30]", "(30, 35]", "(35, 40]", \
                          "(40, 45]"])
age_hist = hist(df$age)
age_freq = age_hist$counts
age_freq_norm = age_freq / length(df$age)

age_df <- data.frame(age_freq, age_freq_norm)
colnames(age_df) <- c("Age frequency", "Relative")
# Not working...
# Create strings for each histogram bin, showing the boundaries as in (20, 25]
for (idx in 2:length(age_hist$breaks))
    bin_bef <- age_hist$breaks[idx-1]
    bin_aft <- age_hist$breaks[idx]
    print(paste( "(", bin_bef, ", ", bin_aft, "]", sep=""))
    age_bins[idx - 1] <- paste( "(", bin_bef, ", ", bin_aft, "]", sep="")

rownames(age_df) <- age_hist$breaks
print()
print("Histogram for age")
print(age_df)

File: Chapter1/S1.4/Python/exercise07.py
import pandas as pd
import matplotlib.pyplot as plt

df = pd.DataFrame()
df["Course"] = 2*["superior"] + 6*["medio"] + 2*["fundamental"]
df["Age"] = [34, 43, 31, 37, 24, 25, 27, 22, 21, 26]
df["Salary"] = [1100.0, 1450.0, 960.0, 960.0] + 3*[600.0] + 3*[450.0]
df["Job Years"] = [5, 8, 6, 8, 3, 2, 5, 2, 3, 3]

# Course variable
course_freq = df.loc[:, "Course"].value_counts()
print("Histogram for course")
print(course_freq)

# Group Age variable
age_bins = pd.cut(df["Age"], (20, 25, 30, 35, 40, 45), \
                  labels=["(20, 25]", "(25, 30]", "(30, 35]", "(35, 40]", \
                          "(40, 45]"])
age_freq = age_bins.value_counts()
age_freq_norm = age_bins.value_counts(normalize=True)

age_df = pd.DataFrame()
age_df["Age frequency"] = age_freq
age_df["Relative"] = age_freq_norm
agd_df = age_df.sort_index()
print()
print("Histogram for age")
print(age_df)

# Group salary
salary_freq = df.Salary.value_counts()
salary_freq_norm = df.Salary.value_counts(normalize=True)
salary_df = pd.DataFrame()
salary_df["Salary frequency"] = salary_freq
salary_df["Relative"] = salary_freq_norm
salary_df = salary_df.sort_index()
print()
print("Histogram for salary")
print(salary_df)

# Analyze how salaries are different considering more than 3 years of job
print()
print("How salaries grow considering years of job")
print(df[["Salary", "Job Years"]].sort_values(by = "Job Years"))

# All plots on the same screen
fig, ((course_plt), (age_plt), (salary_plt)) = plt.subplots(1, 3)
fig.suptitle('Section 1.4 exercise 7')
fig.canvas.set_window_title('Section 1.4 exercise 7')

course_plt.set_title("Course variable")
course_plt.pie(course_freq, labels=course_freq.index, autopct='%1.1f%%')

age_plt.set_title("Age variable")
age_plt.bar(age_df.index.categories, age_df.values[:,0])

salary_plt.set_title("Salary variable")
# I didn't like the salary with the default distance between numbers, so I'm
# converting them to string to have one bar besides the other out of scale.
salary_plt.bar([str(sal) for sal in salary_df.index], salary_df.values[:,0])

plt.show()